godot-core/src/obj/gd.rs

```diff
@@ -545,6 +545,40 @@ impl<T: GodotClass> Gd<T> {
         // Do not increment ref-count; assumed to be return value from FFI.
         sys::ptr_then(object_ptr, |ptr| Gd::from_obj_sys_weak(ptr))
     }
+
+    /// Provides `export_hint` to be used in exports with `Option<Gd<T>>` and `OnEditor<Gd<T>>`.
+    ///
+    /// `Gd<T>` by itself is non-nullable which works badly with Godot editor and might lead to
+    /// unavoidable memory leaks – thus the need to handle `#[export]` with algebraic types.
+    ///
+    /// See also: [#892](https://github.com/godot-rust/gdext/issues/892#issuecomment-2628955463)
+    #[doc(hidden)]
+    pub(crate) fn export_hint() -> PropertyHintInfo
+    where
+        T: Bounds<Exportable = bounds::Yes>,
+    {
+        let hint = if T::inherits::<classes::Resource>() {
+            PropertyHint::RESOURCE_TYPE
+        } else if T::inherits::<classes::Node>() {
+            PropertyHint::NODE_TYPE
+        } else {
+            unreachable!("classes not inheriting from Resource or Node should not be exportable")
+        };
+
+        // Godot does this by default too; the hint is needed when the class is a resource/node,
+        // but doesn't seem to make a difference otherwise.
+        let hint_string = T::class_name().to_gstring();
+
+        PropertyHintInfo { hint, hint_string }
+    }
+
+    #[doc(hidden)]
+    pub(crate) fn as_node_class() -> Option<ClassName>
+    where
+        T: Bounds<Exportable = bounds::Yes>,
+    {
+        T::inherits::<classes::Node>().then(|| T::class_name())
+    }
```

Comment:
`pub(crate)` doesn't need `#[doc(hidden)]`, as it's not part of the public-facing API anyway.

Also, having those as inherent methods with the exact same name as the `Export` trait methods makes it very easy to confuse them for methods overriding the trait. But they would fit well [here](https://github.com/godot-rust/gdext/blob/master/godot-core/src/meta/property_info.rs#L304), as something like:
```rs
fn export_object<T>() -> Self where T: ... { ... }
fn object_as_node_class<T>() -> Option<ClassName> { ... }
```

The other methods there are `pub`, so maybe that's OK? Maybe it was also by accident. But I don't think it hurts much...